src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -27223,6 +27223,50 @@ public class Outer
             compRelease.VerifyEmitDiagnostics();
         }
 
+        [Fact]
+        [WorkItem(50040, "https://github.com/dotnet/roslyn/issues/50040")]
+        public void RaceConditionInAddMembers()
+        {
+            var src = @"
+#nullable enable
+using System;
+using System.Linq.Expressions;
+using System.Threading.Tasks;
+
+var collection = new Collection<Hamster>();
+Hamster h = null!;
+await collection.MethodAsync(entity => entity.Name! == ""bar"", h);
+
+public record Collection<T> where T : Document
+{
+    public Task MethodAsync<TDerived>(Expression<Func<TDerived, bool>> filter, TDerived td) where TDerived : T
+        => throw new NotImplementedException();
+
+    public Task MethodAsync<TDerived2>(Task<TDerived2> filterDefinition, TDerived2 td) where TDerived2 : T
+        => throw new NotImplementedException();
+}
+
+public sealed record HamsterCollection : Collection<Hamster>
+{
+}
+
+public abstract class Document
+{
+}
+
+public sealed class Hamster : Document
+{
+    public string? Name { get; private set; }
+}
+";
+
+            for (int i = 0; i < 10; i++)
```

Comment:
>10 [](start = 32, length = 2)

I am not sure if 10 is big enough number to be confident.

Reply:
The repro rate used to be about 50%, that's why I'm confident with 10.

---
In reply to: [562274980](https://github.com/dotnet/roslyn/pull/50650#discussion_r562274980) [](ancestors = 562274980)

Reply:
> The repro rate used to be about 50%, that's why I'm confident with 10.

I would bump it to 100 to increase the confidence.

---
In reply to: [562314669](https://github.com/dotnet/roslyn/pull/50650#discussion_r562314669) [](ancestors = 562314669,562274980)